[core, lua] Treasure Pool rework

sruon opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [ ] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

- Attaches a CTreasurePoolContainer to CZone, whose job is to manage treasure pools lifecycle.
- Provides a CCharEntity::GetTreasurePool method to ask CZone for the current appropriate treasure pool
- Introduces a Management type for CTreasurePool, indicating if the treasure pool should be managed directly by player/party/alliance or not.
- Moves CTreasurePool ticks processing from zone_entities to the CTreasurePoolContainer
- Introduces a Shared TREASUREPOOLTYPE, set to 36 slots.
- Removes direct PTreasurePool pointer manipulations from CCharEntity/CParty/CAlliance
- Creates an (almost complete) set of Lua bindings to manage treasure pools
- Adds a !getpool debug command, listing target treasure pool members/loot (prints all pools in zone if no target)
- (**Behavior change**) Reworks how Seals are added to pool, to support party-scoped cooldown on seals. On retail, an alliance of 3 parties is 3 distinct sets of cooldown, not a single global one.

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->

## Steps to test these changes

<!-- Clear and detailed steps to test your changes here -->
Still working through testing everything, especially alliance. Will provide full list of tests when able.

sruon committed:
Treasure Pool rework

zach2good commented on src/map/zone.h:
```diff
@@ -678,7 +686,7 @@ class CZone
     void LoadZoneWeather();
 
-    CTreasurePool* m_TreasurePool;
+    std::unique_ptr<CTreasurePoolContainer> m_TreasurePoolContainer;
```
If a treasure pool is just a wrapper around 2x unordered_maps and it is simple to initialize itself - there isn't a need for it to be a unique_ptr, it can just be a regular member of this class

zach2good commented on src/map/zone.h:
```diff
@@ -573,6 +575,12 @@ class CZone
     void   SetLocalVar(const char* var, uint32 val);
     void   ResetLocalVars();
 
+    auto CreateTreasurePool(uint32 poolOwnerId, TREASUREPOOLTYPE poolType) const -> CTreasurePool&;
+    auto GetTreasurePool(CCharEntity* PChar) const -> CTreasurePool&;
+    auto GetTreasurePools() const -> std::unordered_map<uint32, CTreasurePool>&;
+    void ReleaseTreasurePool(CTreasurePool& PTreasurePool) const;
+    auto ReassignTreasurePool(CCharEntity* PPrev, CCharEntity* PNew) -> bool;
+
```
We don't have this written down anywhere, but we're trying to move towards the convention of member functions (even bound ones) being `lowerCamelCase` and free functions (including in namespaces, and bindings) can be `UpperCamelCase`

zach2good commented on src/map/zone.cpp:
```diff
@@ -101,6 +101,8 @@ CZone::CZone(ZONEID ZoneID, REGION_TYPE RegionID, CONTINENT_TYPE ContinentID, ui
     // settings should load first
     LoadZoneSettings();
 
+    m_TreasurePoolContainer = std::make_unique<CTreasurePoolContainer>(m_miscMask & MISC_TREASURE);
```
When I recommended that you make this not a unique_ptr, I didn't see the reading of `m_miscMask` - and now that I'm looking at it: where/when is `m_miscMask` set? Presumably inside `LoadZoneSettings`?

So, no need to change from unique_ptr, this is an item of work for me at some point to pull out all the db information  _before_ we start constructing things

sruon replied:
Correct.

zach2good commented on src/map/treasure_pool_container.h:
```diff
@@ -0,0 +1,56 @@
+===========================================================================
+*/
+
+#ifndef _CTREASUREPOOLCONTAINER_H
```
#pragma once - please

zach2good commented on src/map/treasure_pool_container.cpp:
```diff
@@ -0,0 +1,225 @@
+        if (it->second.MemberCount() < 1)
+        {
+            ShowDebugFmt("Removing empty pool formerly owned by ID {}", it->first);
+            it = m_Pools.erase(it);
```
I tend not to like this pattern of `I am going to iterate with raw iterators and remove things as I go`, can you change this to do similar to what we do in `ZoneEntities`: Where we collect the things we want to remove into a temporary container as we do the first pass, and then remove all the container things. In this case, then you'd flush afterwards.

zach2good commented on src/map/treasure_pool_container.cpp:
```diff
@@ -0,0 +1,225 @@
+// Used when dissolving a party or alliance.
+auto CTreasurePoolContainer::ReleaseTreasurePool(CTreasurePool& PTreasurePool) -> void
+{
+    for (auto it = m_Pools.begin(); it != m_Pools.end();)
```
Same here - as a general rule this isn't the hot-path so we don't need to worry about the mega/micro performance benefits of this type of iteration

zach2good commented on src/map/treasure_pool_container.cpp:
```diff
@@ -0,0 +1,225 @@
+        return false;
+    }
+
+    if (auto node = m_Pools.extract(PPrev->id); !node.empty())
```
Can node be const?

sruon replied:
I don't believe so but while looking at it I realized I don't fully grasp the intricacies of maps and std::move so I might rewrite it in a different way.

zach2good commented on src/map/treasure_pool_container.cpp:
```diff
@@ -0,0 +1,225 @@
+
+// Returns the appropriate treasure pool for PChar in current zone.
+// Priority: Zone wide -> Shared -> Alliance -> Party -> Solo
+auto CTreasurePoolContainer::GetTreasurePool(CCharEntity* PChar) -> CTreasurePool&
```
I _love_ this function, it's such a good abstraction. Nice work

zach2good commented on src/map/treasure_pool.h:
```diff
@@ -26,18 +26,34 @@
+};
+
+// Update xi.treasurePool accordingly when making changes
 enum TREASUREPOOLTYPE
```
While you're here, can you make this: `enum class TreasurePoolType : uint8`, and then the entries will be `Solo`, `Party`, etc.

zach2good commented on src/map/treasure_pool.h:
```diff
@@ -26,18 +26,34 @@
 
 #include <vector>
 
+enum TREASUREPOOLMANAGEMENT
```
Same here, `enum class TreasurePoolManagement : uint8`, `Managed`, `Unmanaged`

zach2good commented on src/map/party.cpp:
```diff
@@ -1047,7 +1013,10 @@ void CParty::SetLeader(const std::string& MemberName)
+        const auto PPrevLeader = dynamic_cast<CCharEntity*>(GetLeader());
+        m_PLeader              = GetMemberByName(MemberName);
+        const auto PNewLeader  = dynamic_cast<CCharEntity*>(m_PLeader);
+
```
As a pattern, if you're doing `dynamic_cast` the immediate next thing should be to check that it's valid. So you'd do this assignment to `m_PLeader` first, and then the two dynamic casts, and then check they're valid

sruon replied:
Changed the flow of that method slightly. PNewLeader will get a static_cast since it's expected in this block that the new leader is a CCharEntity but the PrevLeader may be nullptr (new party creation)

```
void CParty::SetLeader(const std::string& MemberName)
{
    if (m_PartyType == PARTY_PCS)
    {
        uint32 newId = 0;
        int    ret   = _sql->Query(
            "SELECT chars.charid from accounts_sessions JOIN chars ON chars.charid = accounts_sessions.charid WHERE charname = ('%s')", MemberName);

        if (ret != SQL_ERROR && _sql->NumRows() != 0 && _sql->NextRow() == SQL_SUCCESS)
        {
            newId = _sql->GetUIntData(0);
        }
        else
        {
            return;
        }

        _sql->Query("UPDATE accounts_parties SET partyflag = partyflag & ~%d WHERE partyid = %u AND partyflag & %d", ALLIANCE_LEADER | PARTY_LEADER, m_PartyID, PARTY_LEADER);
        _sql->Query("UPDATE accounts_parties SET partyid = %u WHERE partyid = %u", newId, m_PartyID);
        _sql->Query("UPDATE accounts_parties SET allianceid = %u WHERE allianceid = %u", newId, m_PartyID);

        const auto PPrevLeaderEntity = GetLeader();
        m_PLeader                    = GetMemberByName(MemberName);
        const auto PNewLeader        = static_cast<CCharEntity*>(m_PLeader);

        if (this->m_PAlliance && this->m_PAlliance->m_AllianceID == m_PartyID)
        {
            m_PAlliance->m_AllianceID = newId;
        }

        m_PartyID = newId;
        _sql->Query("UPDATE accounts_parties SET partyflag = partyflag | IF(allianceid = partyid, %d, %d) WHERE charid = %u", ALLIANCE_LEADER | PARTY_LEADER, PARTY_LEADER, newId);

        // Passing leader dismisses trusts
        for (auto* PMemberEntity : members)
        {
            if (auto* PMember = dynamic_cast<CCharEntity*>(PMemberEntity))
            {
                PMember->ClearTrusts();

                // In every zone where party members are, reassign the pool to the new leader
                if (PPrevLeaderEntity && PNewLeader)
                {
                    if (const auto PPrevLeader = dynamic_cast<CCharEntity*>(PPrevLeaderEntity))
                    {
                        zoneutils::GetZone(PMember->getZone())->reassignTreasurePool(PPrevLeader, PNewLeader);
                    }
                }
            }
        }
    }
    else
    {
        m_PLeader = members.at(0);
    }
}
```

zach2good commented on src/map/party.cpp:
```diff
@@ -116,6 +126,28 @@ void CParty::DisbandParty(bool playerInitiated)
+        }
+
+        // Force release of treasure pools, otherwise the leader would get reattached to the same pool
+        for (const auto& poolToBeReleased : PTreasurePools)
```
You can do this decomposition in the if block: `if (const auto& [PZone, PTreasurePool] : ...`

zach2good commented on src/map/lua/lua_treasure_pool.h:
```diff
@@ -0,0 +1,49 @@
+*/
+
+#ifndef _LUATREASUREPOOL_H
+#define _LUATREASUREPOOL_H
```
#pragma once

zach2good commented on src/map/lua/lua_treasure_pool.cpp:
```diff
@@ -0,0 +1,190 @@
+    for (const auto& item : m_PLuaTreasurePool->GetItems())
+    {
+        sol::table itemRow   = lua.create_table();
+        itemRow["ID"]        = item.ID;
+        itemRow["SlotID"]    = item.SlotID;
+        itemRow["TimeStamp"] = item.TimeStamp.time_since_epoch().count();
```
Off the top of my head, I don't think we often have `Capitalized` keys when we build Lua tables

zach2good commented on src/map/lua/lua_treasure_pool.cpp:
```diff
@@ -0,0 +1,190 @@
+auto CLuaTreasurePool::getItems() const -> sol::table
+{
+    sol::table result = lua.create_table();
+    int        index  = 1;
```
index should be an unsigned type like size_t, unless it can hold -1, etc.

zach2good commented on src/map/alliance.cpp:
```diff
@@ -237,6 +254,14 @@ void CAlliance::addParty(CParty* party)
 
+    for (auto* PChar : party->members)
+    {
+        if (auto& PTreasurePool = static_cast<CCharEntity*>(PChar)->GetTreasurePool(); PTreasurePool.IsManaged())
```
I know mobs can be in parties, but I don't know if they can have alliances - is it safe to assume entities in here are players and we can `static_cast` them? This isn't the hot path, so we could be checking their objType

zach2good commented on scripts/enum/treasure_pool.lua:
```diff
@@ -0,0 +1,16 @@
+{
+    -- Keep synced with TREASUREPOOLTYPE
+
+    SOLO                     =   1,
```
Spacing here is a bit extreme

zach2good reviewed (approved):
Excellent work - I couldn't find anything logically wrong, so I've been anal about naming conventions, arbitrary C++ things, and some "hot path" things. Do these at your leisure.

🎉

sruon committed:
Treasure Pool rework

sruon commented:
Pushed another commit that addresses review comments and contains bugfixes for alliances.

Most of the tests [are documented here](https://docs.google.com/document/d/1bdE3bKTZtseq3hZi3mLcOkfslbqR-kVVPoYmNqZKH_Q/edit?usp=sharing).

However, when testing across multiple map processes the logic falls apart since there's a strong expectation that party/alliance leaders are set and it's proving too janky to keep everything in sync when party leadership changes.

I'll close the PR and revisit when CParty and CAlliance are reworked.
